@Where parameterization does not work with R2dbc Postgres

### Expected Behavior

Parameterized @Where statements like:

```
    @Where(value = "total_pages > :pages")
    public abstract List<Book> findByTitleStartsWith(String title, int pages);
```
should work and add the additional where statement.

### Actual Behaviour

Query compilation works, but fails at Runtime with the following error:

```java
io.micronaut.transaction.exceptions.TransactionSystemException: Error invoking doInTransaction handler: Binding index 1 when only 1 parameters are expected
	at io.micronaut.data.r2dbc.operations.DefaultR2dbcRepositoryOperations.lambda$withTransaction$15(DefaultR2dbcRepositoryOperations.java:479)
	at reactor.core.publisher.FluxUsingWhen.deriveFluxFromResource(FluxUsingWhen.java:120)
	at reactor.core.publisher.FluxUsingWhen.access$000(FluxUsingWhen.java:54)
	at reactor.core.publisher.FluxUsingWhen$ResourceSubscriber.onNext(FluxUsingWhen.java:193)
	at reactor.core.publisher.Operators$BaseFluxToMonoOperator.completePossiblyEmpty(Operators.java:2034)
	at reactor.core.publisher.MonoHasElements$HasElementsSubscriber.onComplete(MonoHasElements.java:93)
	at reactor.core.publisher.MonoIgnoreElements$IgnoreElementsSubscriber.onComplete(MonoIgnoreElements.java:89)
	at io.r2dbc.postgresql.util.FluxDiscardOnCancel$FluxDiscardOnCancelSubscriber.onComplete(FluxDiscardOnCancel.java:104)
	at reactor.core.publisher.FluxPeek$PeekSubscriber.onComplete(FluxPeek.java:260)
	at reactor.core.publisher.FluxPeek$PeekSubscriber.onComplete(FluxPeek.java:260)
	at reactor.core.publisher.FluxHandle$HandleSubscriber.onComplete(FluxHandle.java:222)
	at io.r2dbc.postgresql.util.FluxDiscardOnCancel$FluxDiscardOnCancelSubscriber.onComplete(FluxDiscardOnCancel.java:104)
	at reactor.core.publisher.FluxContextWrite$ContextWriteSubscriber.onComplete(FluxContextWrite.java:126)
	at reactor.core.publisher.FluxCreate$BaseSink.complete(FluxCreate.java:460)
	at reactor.core.publisher.FluxCreate$BufferAsyncSink.drain(FluxCreate.java:805)
	at reactor.core.publisher.FluxCreate$BufferAsyncSink.complete(FluxCreate.java:753)
	at reactor.core.publisher.FluxCreate$SerializedFluxSink.drainLoop(FluxCreate.java:247)
	at reactor.core.publisher.FluxCreate$SerializedFluxSink.drain(FluxCreate.java:213)
	at reactor.core.publisher.FluxCreate$SerializedFluxSink.complete(FluxCreate.java:204)
	at io.r2dbc.postgresql.client.ReactorNettyClient$Conversation.complete(ReactorNettyClient.java:671)
	at io.r2dbc.postgresql.client.ReactorNettyClient$BackendMessageSubscriber.emit(ReactorNettyClient.java:937)
	at io.r2dbc.postgresql.client.ReactorNettyClient$BackendMessageSubscriber.onNext(ReactorNettyClient.java:813)
	at io.r2dbc.postgresql.client.ReactorNettyClient$BackendMessageSubscriber.onNext(ReactorNettyClient.java:719)
	at reactor.core.publisher.FluxHandle$HandleSubscriber.onNext(FluxHandle.java:128)
	at reactor.core.publisher.FluxPeekFuseable$PeekConditionalSubscriber.onNext(FluxPeekFuseable.java:854)
	at reactor.core.publisher.FluxMap$MapConditionalSubscriber.onNext(FluxMap.java:224)
	at reactor.core.publisher.FluxMap$MapConditionalSubscriber.onNext(FluxMap.java:224)
	at reactor.netty.channel.FluxReceive.drainReceiver(FluxReceive.java:292)
	at reactor.netty.channel.FluxReceive.onInboundNext(FluxReceive.java:401)
	at reactor.netty.channel.ChannelOperations.onInboundNext(ChannelOperations.java:411)
	at reactor.netty.channel.ChannelOperationsHandler.channelRead(ChannelOperationsHandler.java:113)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:336)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:308)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:440)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:788)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:724)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:650)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:562)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: java.lang.IndexOutOfBoundsException: Binding index 1 when only 1 parameters are expected
	at io.r2dbc.postgresql.client.Binding.add(Binding.java:75)
	at io.r2dbc.postgresql.PostgresqlStatement.bind(PostgresqlStatement.java:107)
	at io.r2dbc.postgresql.PostgresqlStatement.bind(PostgresqlStatement.java:58)
	at io.micronaut.data.r2dbc.mapper.R2dbcQueryStatement.setValue(R2dbcQueryStatement.java:144)
	at io.micronaut.data.r2dbc.mapper.R2dbcQueryStatement.setInt(R2dbcQueryStatement.java:222)
	at io.micronaut.data.r2dbc.mapper.R2dbcQueryStatement.setInt(R2dbcQueryStatement.java:39)
	at io.micronaut.data.runtime.mapper.QueryStatement.setDynamic(QueryStatement.java:79)
	at io.micronaut.data.r2dbc.mapper.R2dbcQueryStatement.setDynamic(R2dbcQueryStatement.java:135)
	at io.micronaut.data.r2dbc.mapper.R2dbcQueryStatement.setDynamic(R2dbcQueryStatement.java:39)
	at io.micronaut.data.runtime.operations.internal.sql.AbstractSqlRepositoryOperations.setStatementParameter(AbstractSqlRepositoryOperations.java:238)
	at io.micronaut.data.r2dbc.operations.DefaultR2dbcRepositoryOperations.access$5600(DefaultR2dbcRepositoryOperations.java:133)
	at io.micronaut.data.r2dbc.operations.DefaultR2dbcRepositoryOperations$R2dbcParameterBinder.bindOne(DefaultR2dbcRepositoryOperations.java:1141)
	at io.micronaut.data.runtime.operations.internal.query.DefaultBindableParametersStoredQuery.bindParameter(DefaultBindableParametersStoredQuery.java:168)
	at io.micronaut.data.runtime.operations.internal.query.DefaultBindableParametersStoredQuery.bindParameters(DefaultBindableParametersStoredQuery.java:84)
	at io.micronaut.data.runtime.operations.internal.query.DefaultBindableParametersPreparedQuery.bindParameters(DefaultBindableParametersPreparedQuery.java:82)
	at io.micronaut.data.runtime.operations.internal.query.BindableParametersPreparedQuery.bindParameters(BindableParametersPreparedQuery.java:51)
	at io.micronaut.data.r2dbc.operations.DefaultR2dbcRepositoryOperations$DefaultR2dbcReactiveRepositoryOperations.lambda$findAll$17(DefaultR2dbcRepositoryOperations.java:733)
	at io.micronaut.data.r2dbc.operations.DefaultR2dbcRepositoryOperations.lambda$withTransaction$15(DefaultR2dbcRepositoryOperations.java:474)
	... 51 common frames omitted
```

This appears to be because of [this line in R2dbcRepository](https://github.com/micronaut-projects/micronaut-data/blob/master/data-r2dbc/src/main/java/io/micronaut/data/r2dbc/annotation/R2dbcRepository.java#L49)

when Pattern replacement happens in [AbstractSqlLikeQueryBuilder](https://github.com/micronaut-projects/micronaut-data/blob/master/data-model/src/main/java/io/micronaut/data/model/query/builder/AbstractSqlLikeQueryBuilder.java#L1034),  the `$` character is a special character for group reference, so the replacement fails (it actually substitutes in an empty string).

### Steps To Reproduce

1. Create a Repository with a @Where with a parameter
2. Invoke that method

### Environment Information

MacOS
JDK 17

### Example Application

_No response_

### Version

3.7.2 (happens on lower versions as well)

--- a/data-model/src/main/java/io/micronaut/data/annotation/Where.java
+++ b/data-model/src/main/java/io/micronaut/data/annotation/Where.java
@@ -35,9 +35,10 @@ public @interface Where {
     /**
      * The string value that represents the additional query criterion. For example: {@code enabled = true}
      *
-     * <p>Note that if it may be required to specify the query alias in queries. For example: {@code book_.enabled = true}</p>
+     * <p>Note that if it may be required to specify the query alias in queries. For example: {@code book_.enabled = true}.
+     * `@` can be used as a placeholder for the query's alias</p>
      *
-     * <p>Parameterized variables can be specified using the dollar syntax: {@code book_.enabled = ${enabled}}. In
+     * <p>Parameterized variables can be specified using the dollar syntax: {@code book_.enabled = :enabled}. In
      * this case the parameter must be declared in the method signature a compilation error will occur.</p>
      *
      * <p>Use cases including soft-delete, multi-tenancy etc.</p>

--- a/data-model/src/main/java/io/micronaut/data/model/query/builder/AbstractSqlLikeQueryBuilder.java
+++ b/data-model/src/main/java/io/micronaut/data/model/query/builder/AbstractSqlLikeQueryBuilder.java
@@ -993,9 +993,8 @@ public abstract class AbstractSqlLikeQueryBuilder implements QueryBuilder {
             StringBuilder additionalWhereBuff = buildAdditionalWhereClause(queryState, annotationMetadata);
             if (additionalWhereBuff.length() > 0) {
                 queryClause.append(WHERE_CLAUSE)
-                    .append(OPEN_BRACKET)
-                    .append(additionalWhereBuff.toString())
-                    .append(CLOSE_BRACKET);
+                    .append(OPEN_BRACKET);
+                appendAdditionalWhere(queryClause, queryState, additionalWhereBuff.toString());
             }
         }
     }
@@ -1031,7 +1030,8 @@ public abstract class AbstractSqlLikeQueryBuilder implements QueryBuilder {
             while (matcher.find()) {
                 String name = matcher.group(3);
                 String placeholder = queryState.addAdditionalRequiredParameter(name);
-                matcher.appendReplacement(additionalWhereBuilder, placeholder);
+                //appendReplacement considers $ to be a special character, need to escape it.
+                matcher.appendReplacement(additionalWhereBuilder, Matcher.quoteReplacement(placeholder));
             }
             matcher.appendTail(additionalWhereBuilder);
             additionalWhere = additionalWhereBuilder.toString();

--- a/data-processor/src/test/groovy/io/micronaut/data/model/query/builder/SqlQueryBuilderSpec.groovy
+++ b/data-processor/src/test/groovy/io/micronaut/data/model/query/builder/SqlQueryBuilderSpec.groovy
@@ -102,6 +102,52 @@ interface MyRepository {
         builder.buildInsert(annotationMetadata, entity).query == 'INSERT INTO sale (name,data,quantities,extra_data,data_list) VALUES ($1,to_json($2::json),to_json($3::json),to_json($4::json),to_json($5::json))'
     }
 
+    @Requires({ javaVersion >= 1.8 })
+    void 'test where annotation replacement'() {
+        given:
+        def annotationMetadata = buildTypeAnnotationMetadata('''
+package test;
+import io.micronaut.data.annotation.*;
+import io.micronaut.data.model.query.builder.sql.*;
+import java.lang.annotation.*;
+import io.micronaut.data.jdbc.annotation.*;
+import io.micronaut.context.annotation.*;
+
+@MyAnnotation(dialect = Dialect.POSTGRES)
+@Where("@.name = :name")
+interface MyRepository {
+}
+
+@RepositoryConfiguration(
+        queryBuilder = SqlQueryBuilder.class
+)
+@SqlQueryConfiguration(
+    @SqlQueryConfiguration.DialectConfiguration(
+        dialect = Dialect.POSTGRES,
+        positionalParameterFormat = "$%s",
+        escapeQueries = false
+    )
+)
+@Retention(RetentionPolicy.RUNTIME)
+@Repository
+@interface MyAnnotation {
+    @AliasFor(annotation = Repository.class, member = "dialect")
+    Dialect dialect() default Dialect.ANSI;
+}
+''')
+
+        SqlQueryBuilder builder = new SqlQueryBuilder(annotationMetadata)
+        PersistentEntity entity = PersistentEntity.of(Sale)
+        def queryModel = QueryModel.from(entity)
+
+        expect:
+        builder.dialect == Dialect.POSTGRES
+        builder.buildQuery(annotationMetadata, queryModel).query == 'SELECT sale_.id,sale_.name,sale_.data,sale_.quantities,sale_.extra_data,sale_.data_list FROM sale sale_ WHERE (sale_.name =$1)'
+        builder.buildDelete(annotationMetadata, queryModel).query == 'DELETE  FROM sale  WHERE (name =$1)'
+        builder.buildUpdate(annotationMetadata, queryModel, Arrays.asList("name")).query == 'UPDATE sale SET name=$1 WHERE (name =$2)'
+        builder.buildInsert(annotationMetadata, entity).query == 'INSERT INTO sale (name,data,quantities,extra_data,data_list) VALUES ($1,to_json($2::json),to_json($3::json),to_json($4::json),to_json($5::json))'
+    }
+
 
     void "test encode update with JSON and MySQL"() {
         when:"A update is encoded"

--- a/data-r2dbc/src/test/groovy/io/micronaut/data/r2dbc/postgres/PostgresRepositorySpec.groovy
+++ b/data-r2dbc/src/test/groovy/io/micronaut/data/r2dbc/postgres/PostgresRepositorySpec.groovy
@@ -158,4 +158,16 @@ class PostgresRepositorySpec extends AbstractRepositorySpec implements PostgresT
         return true
     }
 
+    void "test @Where annotation"() {
+        given:
+        setupBooks()
+
+        when:
+        def size = bookRepository.countNativeByTitleWithPagesGreaterThan("The%", 300)
+        def books = bookRepository.findByTitleStartsWith("The", 300)
+
+        then:
+        books.size() == size
+    }
+
 }

--- a/data-r2dbc/src/test/java/io/micronaut/data/r2dbc/postgres/PostgresBookRepository.java
+++ b/data-r2dbc/src/test/java/io/micronaut/data/r2dbc/postgres/PostgresBookRepository.java
@@ -17,6 +17,7 @@ package io.micronaut.data.r2dbc.postgres;
 
 import io.micronaut.data.annotation.Query;
 import io.micronaut.data.annotation.TypeDef;
+import io.micronaut.data.annotation.Where;
 import io.micronaut.data.model.DataType;
 import io.micronaut.data.model.query.builder.sql.Dialect;
 import io.micronaut.data.r2dbc.annotation.R2dbcRepository;
@@ -40,4 +41,10 @@ public abstract class PostgresBookRepository extends BookRepository {
 
     @Query(value = "select * from book where (CASE WHEN exists ( select (:arg0) ) THEN title IN (:arg0) ELSE true END)", nativeQuery = true)
     public abstract List<Book> listNativeBooksNullableArraySearch(@TypeDef(type = DataType.STRING) @Nullable String[] arg0);
+
+    @Where(value = "total_pages > :pages")
+    public abstract List<Book> findByTitleStartsWith(String title, int pages);
+
+    @Query(value = "select count(*) from book b where b.title like :title and b.total_pages > :pages", nativeQuery = true)
+    public abstract int countNativeByTitleWithPagesGreaterThan(String title, int pages);
 }

